File: packages/android-device-server/src/main/java/com/argent/devicecontrol/accessibility/ScreenHash.kt
package com.argent.devicecontrol.accessibility

/**
 * Structural (H) and state (H_text) fingerprints of an accessibility tree
 * (Screen-graph Phase A, design §2.1 / §3).
 *
 * `H` identifies a *screen*: `(className, resourceId, quantized bounds, flags)`
 * per node in DFS order, EXCLUDING text. `H_text` adds `(text, contentDesc)` so
 * it identifies the screen's *state*. `quant` buckets bounds to 1/32 of the
 * screen dims to tolerate small layout jitter; `flags` is the actionability
 * bitmask `clickable|scrollable|editable|checkable|enabled|focused`.
 *
 * The hash is 64-bit FNV-1a over a canonical UTF-8 string. The EXACT same string
 * construction, field order, separators and constants are implemented host-side
 * in TypeScript (`packages/tool-server/src/utils/screen-hash.ts`); on an
 * identical tree the two produce the same hex, so a device hash can be compared
 * to a host-computed one. Edit the two in lockstep.
 */
object ScreenHash {
    // Flag bit positions — MUST match screen-hash.ts.
    const val FLAG_CLICKABLE = 1 shl 0
    const val FLAG_SCROLLABLE = 1 shl 1
    const val FLAG_EDITABLE = 1 shl 2
    const val FLAG_CHECKABLE = 1 shl 3
    const val FLAG_ENABLED = 1 shl 4
    const val FLAG_FOCUSED = 1 shl 5

    /**
     * Password bit (Phase B leftover B1), matching `FLAG_PASSWORD` (1 << 6 = 64)
     * in `packages/tool-server/src/screen-graph/types.ts`. Deliberately NOT part
     * of [flagsOf]: including it would change the structural / state hash of any
     * screen with a password field and break host↔device hash parity (which is
     * bits 0..5 only). It is layered onto the SERIALIZED `flags` of a compact
     * record (see [AxRecords.compact]) purely so the host can redact secrets.
     */
    const val FLAG_PASSWORD = 1 shl 6

    private val FNV_OFFSET = 0xcbf29ce484222325UL
    private val FNV_PRIME = 0x100000001b3UL

    // Field / record separators. Non-printable so they can't occur in a label.
    private const val US = '\u001F'
    private const val RS = '\u001E'

    private val SCROLLING_CONTAINERS = setOf(
        "RecyclerView", "ListView", "ScrollView", "HorizontalScrollView"
    )

    /**
     * Window-decor resource-ids that appear/disappear independently of the screen
     * (the status-bar / nav-bar backgrounds). Excluded from `H_id` so the same
     * screen hashes identically whether or not the decor node was captured
     * (screen-graph Phase D §1 — "exclude the status bar window").
     */
    private val SYSTEM_RIDS = setOf("statusBarBackground", "navigationBarBackground")

    /**
     * Text that is volatile between otherwise-identical screen states — a clock, a
     * battery/signal percentage, a size, a date, a bare counter. Kept in lockstep
     * with `VOLATILE_TEXT` in `packages/tool-server/src/screen-graph/plan.ts` and
     * the host `screen-hash.ts` twin. An identity title matching this is dropped
     * from `H_id` so a live counter cannot split a screen from itself.
     */
    private val VOLATILE_TEXT = Regex(
        "^\\s*(?:\\d{1,3}\\s*%|\\d{1,2}:\\d{2}(?:\\s*[ap]m)?|" +
            "\\d[\\d.,]*\\s*(?:%|min|hr|hrs|h|GB|MB|KB|B)?|" +
            "[A-Z][a-z]{2}\\s+\\d{1,2}|\\d{1,2}/\\d{1,2}(?:/\\d{2,4})?)\\s*$",
        RegexOption.IGNORE_CASE
    )

    fun isVolatileText(text: String): Boolean = VOLATILE_TEXT.matches(text.trim())

    fun flagsOf(n: AxNode): Int {
        var f = 0
        if (n.clickable) f = f or FLAG_CLICKABLE
        if (n.scrollable) f = f or FLAG_SCROLLABLE
        if (n.editable) f = f or FLAG_EDITABLE
        if (n.checkable) f = f or FLAG_CHECKABLE
        if (n.enabled) f = f or FLAG_ENABLED
        if (n.focused) f = f or FLAG_FOCUSED
        return f
    }

    /** A scrolling container is hashed by container + its first child's class seq only. */
    fun isScrollingContainer(n: AxNode): Boolean {
        if (n.scrollable) return true
        return n.className in SCROLLING_CONTAINERS || n.className.startsWith("ViewPager")
    }

    private fun quant(v: Int, dim: Int): Int {
        if (dim <= 0) return 0
        return Math.floorDiv(v.toLong() * 32L, dim.toLong()).toInt()
    }

    /** H — structural, text-excluded, recycler rule applied. */
    fun structural(roots: List<AxNode>, screenW: Int, screenH: Int): String {
        val sb = StringBuilder()
        for (r in roots) appendStructural(r, screenW, screenH, sb)
        return fnv1a(sb.toString())
    }

    private fun appendStructural(n: AxNode, w: Int, h: Int, sb: StringBuilder) {
        sb.append(n.className).append(US)
            .append(n.resourceId).append(US)
            .append(quant(n.x1, w)).append(',').append(quant(n.y1, h)).append(',')
            .append(quant(n.x2, w)).append(',').append(quant(n.y2, h)).append(US)
            .append(flagsOf(n)).append(RS)
        if (isScrollingContainer(n)) {
            // Recycler rule (design §6): container node + the class sequence of
            // its FIRST child subtree only, so item count / content does not
            // change H. Remaining children are skipped for H.
            val first = n.children.firstOrNull()
            if (first != null) appendClassSeq(first, sb)
        } else {
            for (c in n.children) appendStructural(c, w, h, sb)
        }
    }

    private fun appendClassSeq(n: AxNode, sb: StringBuilder) {
        sb.append(n.className).append(RS)
        for (c in n.children) appendClassSeq(c, sb)
    }

    /** H_text — structural + (text, contentDesc), ALL children, no recycler rule. */
    fun state(roots: List<AxNode>, screenW: Int, screenH: Int): String {
        val sb = StringBuilder()
        for (r in roots) appendState(r, screenW, screenH, sb)
        return fnv1a(sb.toString())
    }

    private fun appendState(n: AxNode, w: Int, h: Int, sb: StringBuilder) {
        sb.append(n.className).append(US)
            .append(n.resourceId).append(US)
            .append(quant(n.x1, w)).append(',').append(quant(n.y1, h)).append(',')
            .append(quant(n.x2, w)).append(',').append(quant(n.y2, h)).append(US)
            .append(flagsOf(n)).append(US)
            .append(n.text).append(US)
            .append(n.contentDesc).append(RS)
        for (c in n.children) appendState(c, w, h, sb)
    }

    /* ---------------------------------------------------------------------- */
    /* H_id — screen IDENTITY (screen-graph Phase D §1)                         */
    /* ---------------------------------------------------------------------- */

    /** A container whose id marks it as a toolbar / action-bar / app-bar. */
    fun isToolbarContainer(resourceId: String): Boolean {
        val r = resourceId.lowercase()
        return r.contains("toolbar") || r == "action_bar" || r.contains("app_bar")
    }

    /**
     * A node whose text is the screen's IDENTITY title: a collapsing-toolbar /
     * action-bar / dialog title, or a bare `title` sitting under a toolbar. The
     * Settings homepage's oversized `homepage_title` is deliberately NOT matched:
     * it collapses out of the tree on scroll (run 33806639520 roots 77a189ce vs
     * 299378e0), so keying identity on it would split one screen in two. Search
     * hints are excluded.
     */
    fun isIdentityTitle(resourceId: String, ancestorIsToolbar: Boolean): Boolean {
        val r = resourceId.lowercase()
        if (r.contains("search")) return false
        if (r.contains("collapsing_toolbar")) return true
        if (r == "action_bar_title" || r == "toolbar_title" || r == "actionbar_title" || r == "alerttitle") return true
        if (r == "title" && ancestorIsToolbar) return true
        return false
    }

    /**
     * H_id — the SCREEN identity used for graph node keys and routing (design D §1).
     * Unlike [structural] (`H`, which is scroll- and focus-sensitive: it appends a
     * scrolling container's first-child class sequence, quantised bounds and the
     * focus flag, so every Settings detail screen collapses onto one hash), `H_id`
     * is built to be STABLE across scroll and focus and DISTINCT across sibling
     * screens:
     *
     *   package
     *   + the texts of identity nodes (collapsing-toolbar / action-bar / dialog
     *     titles, and a `title` under a toolbar), volatile text dropped;
     *   + the resource-id MULTISET of non-scrollable subtrees, order-free;
     *   + for every scrolling container, `class#resourceId` ONLY — never its child
     *     sequence, never bounds.
     *
     * No bounds, no focus flags, no list content enter the hash. `H` and `H_text`
     * are unchanged (diff / awaitChange still use them). Same 64-bit FNV-1a and
     * US/RS separators as the other two, mirrored host-side in
     * `packages/tool-server/src/utils/screen-hash.ts` — edit the two in lockstep.
     */
    fun identity(roots: List<AxNode>, packageName: String): String {
        val titles = sortedSetOf<String>()
        val nonScrollRids = ArrayList<String>()
        val scTokens = sortedSetOf<String>()

        fun collectTitle(n: AxNode) {
            val t = (if (n.text.isNotEmpty()) n.text else n.contentDesc).trim()
            if (t.isNotEmpty() && !isVolatileText(t)) titles.add(t)
        }
        fun scanTitles(n: AxNode, ancestorIsToolbar: Boolean) {
            if (isIdentityTitle(n.resourceId, ancestorIsToolbar)) collectTitle(n)
            val childToolbar = ancestorIsToolbar || isToolbarContainer(n.resourceId)
            for (c in n.children) scanTitles(c, childToolbar)
        }
        fun scanRids(n: AxNode, insideScroll: Boolean) {
            if (isScrollingContainer(n)) {
                scTokens.add("SC:${n.className}#${n.resourceId}")
                for (c in n.children) scanRids(c, true)
            } else {
                if (!insideScroll && n.resourceId.isNotEmpty() && n.resourceId !in SYSTEM_RIDS) {
                    nonScrollRids.add(n.resourceId)
                }
                for (c in n.children) scanRids(c, insideScroll)
            }
        }
        for (r in roots) {
            scanTitles(r, isToolbarContainer(r.resourceId))
            scanRids(r, false)
        }
        nonScrollRids.sort()
        val sb = StringBuilder()
        sb.append(packageName).append(US)
            .append("ID:").append(titles.joinToString("|")).append(US)
            .append("RID:").append(nonScrollRids.joinToString(",")).append(US)
            .append("SC:").append(scTokens.joinToString(","))
        return fnv1a(sb.toString())
    }

    private fun fnv1a(s: String): String {
        var hash = FNV_OFFSET
        for (b in s.toByteArray(Charsets.UTF_8)) {
            hash = hash xor (b.toInt() and 0xFF).toULong()
            hash *= FNV_PRIME
        }
        return java.lang.Long.toUnsignedString(hash.toLong(), 16).padStart(16, '0')
    }
}

File: packages/tool-server/src/utils/screen-hash.ts
/**
 * Host-side structural (H) and state (H_text) fingerprints of an accessibility
 * tree (Screen-graph Phase A, design §2.1 / §3).
 *
 * This is the TypeScript twin of the on-device Kotlin `ScreenHash`
 * (`packages/android-device-server/.../accessibility/ScreenHash.kt`): identical
 * flag bits, quantization, field order, US/RS separators and
 * 64-bit FNV-1a constants, so the two produce the same hex for the same tree.
 *
 * On the Android open path the DEVICE is the authority for H/H_text (it hashes
 * the real hierarchical tree and returns the hex in `getState`/`query`/`diff`/
 * `awaitChange`); this module is used for host-side hashing where a hierarchical
 * tree is available, and is the unit-tested reference for determinism and the
 * recycler rule. The open server serves a FLATTENED tree to the host, so the two
 * are not machine-cross-verified on that path — they agree by shared construction.
 */

/** A node shaped for hashing. All string fields default to "" when absent. */
export interface HashNode {
  class?: string;
  /** Stripped resource id (the part after "id/"). */
  id?: string;
  text?: string;
  cd?: string;
  bounds: { x1: number; y1: number; x2: number; y2: number };
  clickable?: boolean;
  scrollable?: boolean;
  editable?: boolean;
  checkable?: boolean;
  enabled?: boolean;
  focused?: boolean;
  children?: HashNode[];
}

// Flag bit positions — MUST match ScreenHash.kt.
export const FLAG_CLICKABLE = 1 << 0;
export const FLAG_SCROLLABLE = 1 << 1;
export const FLAG_EDITABLE = 1 << 2;
export const FLAG_CHECKABLE = 1 << 3;
export const FLAG_ENABLED = 1 << 4;
export const FLAG_FOCUSED = 1 << 5;

const US = String.fromCharCode(0x1f); // unit separator between fields
const RS = String.fromCharCode(0x1e); // record separator between nodes

const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const MASK64 = 0xffffffffffffffffn;

const SCROLLING_CONTAINERS = new Set(["RecyclerView", "ListView", "ScrollView", "HorizontalScrollView"]);

/**
 * Window-decor resource-ids excluded from `H_id` (status/nav bar backgrounds),
 * so the same screen hashes identically whether or not the decor was captured.
 * MUST match `ScreenHash.SYSTEM_RIDS` (Kotlin).
 */
const SYSTEM_RIDS = new Set(["statusBarBackground", "navigationBarBackground"]);

/**
 * Volatile text dropped from `H_id` identity titles — a clock, percentage, size,
 * date or bare counter. MUST stay in lockstep with `ScreenHash.VOLATILE_TEXT`
 * (Kotlin) and `VOLATILE_TEXT` in `screen-graph/plan.ts`.
 */
const HID_VOLATILE_TEXT =
  /^\s*(?:\d{1,3}\s*%|\d{1,2}:\d{2}(?:\s*[ap]m)?|\d[\d.,]*\s*(?:%|min|hr|hrs|h|GB|MB|KB|B)?|[A-Z][a-z]{2}\s+\d{1,2}|\d{1,2}\/\d{1,2}(?:\/\d{2,4})?)\s*$/i;

/** True when `text` is purely volatile content (mirror of `ScreenHash.isVolatileText`). */
export function isVolatileText(text: string): boolean {
  return HID_VOLATILE_TEXT.test(text.trim());
}

/** A container whose id marks it a toolbar / action-bar / app-bar (Kotlin twin). */
export function isToolbarContainer(resourceId: string): boolean {
  const r = resourceId.toLowerCase();
  return r.includes("toolbar") || r === "action_bar" || r.includes("app_bar");
}

/**
 * A node whose text is the screen's IDENTITY title (Kotlin twin). The Settings
 * homepage `homepage_title` is deliberately NOT matched: it collapses out of the
 * tree on scroll, so keying identity on it splits one screen in two.
 */
export function isIdentityTitle(resourceId: string, ancestorIsToolbar: boolean): boolean {
  const r = resourceId.toLowerCase();
  if (r.includes("search")) return false;
  if (r.includes("collapsing_toolbar")) return true;
  if (r === "action_bar_title" || r === "toolbar_title" || r === "actionbar_title" || r === "alerttitle") return true;
  if (r === "title" && ancestorIsToolbar) return true;
  return false;
}

export function flagsOf(n: HashNode): number {
  let f = 0;
  if (n.clickable) f |= FLAG_CLICKABLE;
  if (n.scrollable) f |= FLAG_SCROLLABLE;
  if (n.editable) f |= FLAG_EDITABLE;
  if (n.checkable) f |= FLAG_CHECKABLE;
  if (n.enabled) f |= FLAG_ENABLED;
  if (n.focused) f |= FLAG_FOCUSED;
  return f;
}

export function isScrollingContainer(n: HashNode): boolean {
  if (n.scrollable) return true;
  const c = n.class ?? "";
  return SCROLLING_CONTAINERS.has(c) || c.startsWith("ViewPager");
}

function quant(v: number, dim: number): number {
  if (dim <= 0) return 0;
  return Math.floor((v * 32) / dim);
}

function fnv1a(s: string): string {
  let h = FNV_OFFSET;
  const bytes = Buffer.from(s, "utf8");
  for (const b of bytes) {
    h ^= BigInt(b);
    h = (h * FNV_PRIME) & MASK64;
  }
  return h.toString(16).padStart(16, "0");
}

/** H — structural, text-excluded, recycler rule applied. */
export function structuralHash(roots: HashNode[], screenW: number, screenH: number): string {
  let out = "";
  const append = (n: HashNode): void => {
    out +=
      (n.class ?? "") + US +
      (n.id ?? "") + US +
      quant(n.bounds.x1, screenW) + "," + quant(n.bounds.y1, screenH) + "," +
      quant(n.bounds.x2, screenW) + "," + quant(n.bounds.y2, screenH) + US +
      flagsOf(n) + RS;
    const children = n.children ?? [];
    if (isScrollingContainer(n)) {
      // Recycler rule (design §6): container + the class sequence of its FIRST
      // child subtree only, so item count / content does not change H.
      if (children.length > 0) appendClassSeq(children[0]!);
    } else {
      for (const c of children) append(c);
    }
  };
  const appendClassSeq = (n: HashNode): void => {
    out += (n.class ?? "") + RS;
    for (const c of n.children ?? []) appendClassSeq(c);
  };
  for (const r of roots) append(r);
  return fnv1a(out);
}

/**
 * H_id — the SCREEN identity used for graph node keys and routing (design D §1),
 * the host twin of `ScreenHash.identity` (Kotlin). Unlike `H`/`structuralHash`
 * (scroll- and focus-sensitive), `H_id` is stable across scroll/focus and
 * distinct across sibling screens:
 *   package
 *   + identity-node texts (collapsing-toolbar / action-bar / dialog titles, and a
 *     `title` under a toolbar), volatile text dropped;
 *   + the resource-id MULTISET of non-scrollable subtrees, order-free;
 *   + `class#id` ONLY for every scrolling container — never its child sequence,
 *     never bounds.
 * No bounds, no focus flags, no list content. Same FNV-1a + US/RS as `H`/`H_text`.
 */
export function identityHash(roots: HashNode[], packageName: string): string {
  const titles = new Set<string>();
  const nonScrollRids: string[] = [];
  const scTokens = new Set<string>();

  const collectTitle = (n: HashNode): void => {
    const t = ((n.text ?? "") !== "" ? n.text! : n.cd ?? "").trim();
    if (t !== "" && !isVolatileText(t)) titles.add(t);
  };
  const scanTitles = (n: HashNode, ancestorIsToolbar: boolean): void => {
    if (isIdentityTitle(n.id ?? "", ancestorIsToolbar)) collectTitle(n);
    const childToolbar = ancestorIsToolbar || isToolbarContainer(n.id ?? "");
    for (const c of n.children ?? []) scanTitles(c, childToolbar);
  };
  const scanRids = (n: HashNode, insideScroll: boolean): void => {
    if (isScrollingContainer(n)) {
      scTokens.add(`SC:${n.class ?? ""}#${n.id ?? ""}`);
      for (const c of n.children ?? []) scanRids(c, true);
    } else {
      const id = n.id ?? "";
      if (!insideScroll && id !== "" && !SYSTEM_RIDS.has(id)) nonScrollRids.push(id);
      for (const c of n.children ?? []) scanRids(c, insideScroll);
    }
  };
  for (const r of roots) {
    scanTitles(r, isToolbarContainer(r.id ?? ""));
    scanRids(r, false);
  }
  nonScrollRids.sort();
  const out =
    packageName + US +
    "ID:" + [...titles].sort().join("|") + US +
    "RID:" + nonScrollRids.join(",") + US +
    "SC:" + [...scTokens].sort().join(",");
  return fnv1a(out);
}

/** H_text — structural + (text, contentDesc), ALL children, no recycler rule. */
export function stateHash(roots: HashNode[], screenW: number, screenH: number): string {
  let out = "";
  const append = (n: HashNode): void => {
    out +=
      (n.class ?? "") + US +
      (n.id ?? "") + US +
      quant(n.bounds.x1, screenW) + "," + quant(n.bounds.y1, screenH) + "," +
      quant(n.bounds.x2, screenW) + "," + quant(n.bounds.y2, screenH) + US +
      flagsOf(n) + US +
      (n.text ?? "") + US +
      (n.cd ?? "") + RS;
    for (const c of n.children ?? []) append(c);
  };
  for (const r of roots) append(r);
  return fnv1a(out);
}
